### ISSUE TITLE
Define __cuda_array_interface__ equivalent for ROC GPUs? 

### ISSUE THREAD

**Issue by leofang** [numba/numba#4566]
## Feature request

I am trying to think one step ahead, but perhaps I am not knowledgable enough about the current status.

It seems CuPy is about to support ROCm (experimentally, see cupy/cupy#1094). It'd be nice to have a `__cuda_array_interface__` equivalent for ROC GPU arrays, say `__roc_array_interface__` or, more generally, a unified standard `__gpu_array_interface__` applicable for both sides.

Currently I do not have access to AMD GPUs, but I'm interested in seeing this happening to prepare for future access.

cc: @okuta @mitmul

**Comment by leofang** [comment]
A general interface like `__gpu_array_interface__` has the additional benefit that backends other than CUDA and ROC can also be supported, such as OpenCL. AFAIK, currently there's ClPy (https://github.com/fixstars/clpy, a CuPy fork that uses OpenCL) and PyOpenCL (https://github.com/inducer/pyopencl). But I know almost nothing about OpenCL, so I am unable to say more.

**Comment by seibert** [comment]
Long term, I really do hope that HMM support in Linux will obsolete many of these concerns:

https://www.kernel.org/doc/html/v4.18/vm/hmm.html

(That doesn't help us today, but just wanted to throw that out there.)

**Comment by leofang** [comment]
> CUDA has a unified memory space, so we don't need to track which physical device the memory is stored on (and the CUDA API has a way to determine which device a given pointer is on), but I don't know if that is true for other accelerator APIs.

This is a good point. Looking at [the HIP code](https://github.com/ROCm-Developer-Tools/HIP/blob/6031b5ab7665e4aa21367f372bc0e5c1e434992f/hipify-clang/src/CUDA2HIP_Driver_API_functions.cpp#L322-L328) I don't think there's an equivalence of `cuPointerGetAtrribute()` for now. 

> Long term, I really do hope that HMM support in Linux will obsolete many of these concerns
> (That doesn't help us today, but just wanted to throw that out there.)

I wasn't aware of HMM, looks interesting. Thanks. I am also throwing things out here and see if some ideas can be formed along the way.

**Comment by leofang** [comment]
Just a quick update: while HIP does not have an equivalent `cuPointerGetAtrribute` in its driver API, it does have `hipPointerGetAttributes` in its runtime API, which is equivalent to `cudaPointerGetAttributes` from CUDA runtime. I also think HIP has a unified memory space like in CUDA.

**Comment by leofang** [comment]
@seibert @gmarkall I would like to revisit this issue in the near future. I recently started working on improving CuPy's support for ROCm (you can search my id + "ROCm" there), just to evaluate the capability and limitation on the AMD side. I've been able to make CuPy run on a Radeon VII for several features, but more extensive efforts are required since its current status is at best experimental. 

As a small "milestone", I'd like to be able to make CuPy and Numba interoperable (again). What do you think if I add `__rocm_array_interface__` as an alias of `__cuda_array_interface__`? On the CuPy side, I see the tests for `__cuda_array_interface__` passed without problem on AMD GPU, so I think this is the most promising approach with minimized effort.

**Comment by gmarkall** [comment]
> As a small "milestone", I'd like to be able to make CuPy and Numba interoperable (again). What do you think if I add `__rocm_array_interface__` as an alias of `__cuda_array_interface__`? On the CuPy side, I see the tests for `__cuda_array_interface__` passed without problem on AMD GPU, so I think this is the most promising approach with minimized effort.

I don't have much expertise on the ROC side, but I have noted that it's quite similar to the CUDA target. I'm not sure what you mean by making `__rocm_array_interface__` an alias of `__cuda_array_interface__` though - I can't see that it would make sense for anything CUDA-related in Numba to have a `__rocm_array_interface__`. The CUDA and ROC device arrays are defined separately in `numba.cuda.cudadrv.devicearray` and `numba.roc.hsadrv.devicearray`, and I think the only common infrastructure they share is from `numba.misc.dummyarray`, which contains things related to shapes, strides, extents, etc.

**Comment by leofang** [comment]
Perhaps I worded it poorly, Graham 😅 By "making an alias" I mean they obey the same protocol and expose the same fields, since they behave rather similarly as you noticed. So this is what I'd expect from CuPy's point of view:
```python
# pseudo code for cupy.ndarray
class ndarray:
    ...
    
    @property
    def __cuda_array_interface__(self):
        cdef dict cai = {}
        # ...prepare the dict...
        return cai

    @property
    def __rocm_array_interface__(self):
        return self.__cuda_array_interface__  # just an alias!
```
and perhaps `numba.misc.dummyarray` could behave similarly? (I need to find time to look into Numba's implementation, so right now I'm just throwing out a bunch of questions 😅)

With this, I can then make both the Numba CUDA and ROCm kernels accept a `cupy.ndarray` like this:
```python
import cupy
from numba import cuda

@cuda.jit
def f(a):
    pass

# on NVIDIA GPU (already doable today)
x = cupy.zeros(10)
f[1, 16](x)
```
and this:
```python
import cupy
from numba import roc

@roc.jit
def f(a):
    pass

# on AMD GPU (my goal)
x = cupy.zeros(10)
f[1, 16](x)
```

**Comment by gmarkall** [comment]
> Perhaps I worded it poorly, Graham sweat_smile By "making an alias" I mean they obey the same protocol and expose the same fields, since they behave rather similarly as you noticed.

OK, I understand now :-)

> So this is what I'd expect from CuPy's point of view:
> 
> ...
>
> and perhaps `numba.misc.dummyarray` could behave similarly? (I need to find time to look into Numba's implementation, so right now I'm just throwing out a bunch of questions sweat_smile)

The `dummyarray` code doesn't do much, and only really knows about shapes and strides of arrays - the `__rocm_array_interface__` would instead have to go into the ROC-specific code - probably `numba.roc.hsadrv.devicearray`, which is the ROC counterpart of the CUDA device arrays where `__cuda_array_interface__` is defined.

**Comment by leofang** [comment]
Thanks @gmarkall. Unfortunately since I can't run Numba on my AMD GPU, I cannot proceed and make any changes for the ROCm target until #6196 is fixed by team 😢

**Comment by leofang** [comment]
A little update on naively using `__cuda_array_interface__` on HIP: they don't support stream 1 or 2, so passing these streams will segfault: https://github.com/cupy/cupy/pull/4458 😂 We ended up disabling it (in another PR), and will revisit it once the present issue is resolved.

Another thing is there're discussion on DLPack in the Python Array API Consortium, which aims to support CPU/CUDA/HIP/etc. For example, we've been discussing how to ensure async operations (like what we've done with CAI v3). A few relevant pointers (the discussions are scattered in a few repos...):
- https://github.com/data-apis/array-api/pull/106
- https://github.com/data-apis/consortium-feedback/issues/1
- https://github.com/dmlc/dlpack/issues/57

**Comment by leofang** [comment]
Closing this request as ROCm is no longer supported (#6991).
